samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -705,43 +714,42 @@ public void run() {
     this.sideInputSystemConsumers.start();
 
     // create a thread for sideInput reads
-    Thread readSideInputs = new Thread(() -> {
-        while (!shutDownSideInputRead) {
-          IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
-          if (envelope != null) {
-
-            if (!envelope.isEndOfStream())
-              sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
+    Thread readSideInputsThread = new Thread(() -> {
+        try {
+          while (!shutDownSideInputRead) {
+            IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
+            if (envelope != null) {
 
-            checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
-                SystemStreamMetadata.OffsetType.NEWEST, envelope.isEndOfStream());
+              if (!envelope.isEndOfStream())
+                sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
 
-          } else {
-            LOG.trace("No incoming message was available");
+              checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
+                  SystemStreamMetadata.OffsetType.NEWEST, envelope.isEndOfStream());
+            } else {
+              LOG.trace("No incoming message was available");
+            }
           }
+        } catch (Exception e) {
+          LOG.error("Exception in reading side inputs", e);
+          sideInputException = Optional.of(e);
         }
       });
-
-    readSideInputs.setDaemon(true);
-    readSideInputs.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
-      @Override
-      public void uncaughtException(Thread t, Throwable e) {
-        sideInputException = Optional.of(e);
-        sideInputsCaughtUp.countDown();
-      }
-    });
+    readSideInputsThread.setName(SIDEINPUTS_READ_THREAD_NAME);
+    readSideInputsThread.setDaemon(true);
 
     try {
-      readSideInputs.start();
-      // Make the main thread wait until all sideInputs have been caughtup or thrown an exception
-      this.sideInputsCaughtUp.await();
+      readSideInputsThread.start();
+
+      // Make the main thread wait until all sideInputs have been caughtup or an exception was thrown
+      while (!sideInputException.isPresent() && !this.sideInputsCaughtUp.await(SIDE_INPUT_READ_THREAD_TIMEOUT.getSeconds(), TimeUnit.SECONDS)) {
```

Comment:
don't do boiler-plate conversions back and forth - store durations in seconds to begin with